coveo-functools/coveo_functools/flex/serializer.py

```diff
@@ -0,0 +1,65 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from importlib import import_module
+from inspect import isclass
+from typing import Any, Dict
+
+from coveo_functools.annotations import find_annotations
+from coveo_functools.flex.types import TypeHint
+
+
+@dataclass
+class SerializationMetadata:
+    module_name: str
+    class_name: str
+    additional_metadata: Dict[str, SerializationMetadata] = field(default_factory=dict)
+
+    @classmethod
+    def from_instance(cls, instance: Any) -> SerializationMetadata:
+        """Generates serialization metadata out of an instance."""
+        if instance is None:
+            raise Exception(":travolta-meme:")
+
+        if isclass(instance) or not hasattr(instance, "__class__"):
+            raise Exception("Can only reliably serialize from instances.")
+
+        actual_type = instance.__class__
+
+        additional_metadata: Dict[str, SerializationMetadata] = {}
+
+        if isinstance(instance, list):
+            # the additional metadata will be a map of the index to that object's type (order is important)
+            additional_metadata = {
+                str(idx).zfill(6): SerializationMetadata.from_instance(obj)

```

Comment:
You probably should verify that `len(instance) < 1e6`, otherwise, you could have strings with 7 characters here (instead than the expected 6), right?

Reply:
yeah I guess I should fix that. it's dumb, i used an int first, but mypy was giving me trouble with the str|int so I worked around it. But I think you're right, the extra work here will be a peace of mind.